libbpf-tools/lsan.c

```diff
@@ -0,0 +1,1331 @@
+		if (err < 0) {
+			return err;
+		}
+		item = (struct lsan_hash_t*)malloc(sizeof(struct lsan_hash_t));
+		ON_MEM_FAILURE(item);
+		item->size = val.size;
+		item->stack_id = val.stack_id;
+		item->tag = val.tag;
+		item->id = next_key;
+		HASH_ADD(hh, copied, id, sizeof(uptr), item);
+		cvector_push_back(key_table, next_key);
+		HASH_ITER(hh, uncertain, curr, next) {
+			if (curr->id <= next_key && next_key < curr->end) {
+				curr->is_heap = true;
+				break;
+			}
+		}
+		lookup_key = next_key;
+	}
+	qsort(key_table, cvector_size(key_table), sizeof(uptr), compare);
+	return 0;
+}
+
+static void save_roots()
+{
+	char file_name[FILENAME_MAX] = {0, };
+	char line[LINE_MAX] = {0, };
+	FILE *fp = NULL;
+	char* v[MAPS_COLUMN_MAX] = {NULL, };
+	int i = 0;
+	char *ptr = NULL;
+	uptr begin = 0;
+	uptr end = 0;
+	const int hex = 16;
+	struct root_region_t *item = NULL;
+	struct root_region_t *val = NULL;
+
+	snprintf(file_name, sizeof(file_name), "/proc/%d/maps", env.pid);
+	fp = fopen(file_name, "rt");
+	if (fp == NULL) {
+		p_err("Failed to open : %s", file_name);
+		return;
+	}
+	while (fgets(line, sizeof(line), fp) != NULL) {
+		i = 0;
+		ptr = strtok(line, " ");
+		while (ptr != NULL) {
+			v[i] = strdup(ptr);
+			ON_MEM_FAILURE(v[i]);
+			++i;
+			ptr = strtok(NULL, " ");
+		}
+		/* root should have rw permission */
+		if (0 == strncmp(rw_permission, v[MAPS_PERMISSIONS], strlen(rw_permission))) {
+			ptr = strtok(v[MAPS_ADDRESS], "-");
+			begin = strtoull(ptr, NULL, hex);
+			ptr = strtok(NULL, "-");
+			end = strtoull(ptr, NULL, hex);
+			item = (struct root_region_t*)malloc(sizeof(struct root_region_t));
+			ON_MEM_FAILURE(item);
+			item->end = end;
+			item->id = begin;
+			if ('\n' == v[MAPS_PATH][0]) {
+				/* v[MAPS_PATH] value is "\n" if path name does not exist */
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if (0 == strncmp(heap_str, v[MAPS_PATH], strlen(heap_str))) {
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = true;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if ('[' != v[MAPS_PATH][0] ||
+					0 == strncmp(stack_str, v[MAPS_PATH], strlen(stack_str))) {
+				/* Not including '[' starting path except "[stack]" */
+				HASH_FIND(hh, certain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, certain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			}
+		}
+		for (i = 0; i < MAPS_COLUMN_MAX; ++i) {
+			free(v[i]);
+		}
+	}
+	fclose(fp);
+}
+
+static void process_roots(struct root_region_t *roots) {
+	struct root_region_t *curr = NULL;
+	struct root_region_t *next = NULL;
+
+	HASH_ITER(hh, roots, curr, next) {
+		if (!(curr->is_heap)) {
```

Comment:
Here(function `process_roots`), it targets only memory pointers which are not in the heap.
So if you have a pointer inside the heap, it will not be used in this function.
That pointer inside the heap could be scanned if it is reachable directly from the roots,
and the pointer inside the heap cannot be scanned if it is not reachable directly from the roots.
The pointer inside the heap will be scanned (on `flood_fill_tag` function), if there is a path from the root.

Leak sanitizer algorithm is doing `one-depth DFS` for roots and `BFS` for the `DFS` results.
It uses any value as `node` and pointer as `edge`. (Pointer is a `node` as well as `edge`)
`process_roots` is a function do `one-depth DFS` and `flood_fill_tag` is a function do `DFS`